File: matching-engine/src/main/java/com/titanx/engine/OrderBook.java
package com.titanx.engine;

import com.titanx.model.*;

import java.math.BigDecimal;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * ============================================================
 * TitanX OrderBook — Phase 1 Core Data Structure
 * ============================================================
 *
 * Design:
 *   - BIDS:  TreeMap<price, PriorityQueue<Order>> — DESCENDING key order
 *             → best bid is the HIGHEST price (first entry)
 *   - ASKS:  TreeMap<price, PriorityQueue<Order>> — ASCENDING key order
 *             → best ask is the LOWEST price (first entry)
 *
 * Within each price level, orders are in a PriorityQueue sorted by
 * arrival time (price-time priority — standard exchange semantics).
 *
 * Interview talking point:
 *   "I chose TreeMap over HashMap because matching requires sorted
 *    iteration — you must walk price levels in order. TreeMap gives
 *    O(log n) insert/delete and O(1) first/last, which is the right
 *    trade-off for a matching engine."
 *
 * Thread safety:
 *   This class is NOT thread-safe. External locking (per-symbol
 *   ReentrantReadWriteLock in MatchingEngine) is the caller's
 *   responsibility. This keeps the hot path clean.
 */
public class OrderBook {

    private final String symbol;

    // Best bid = highest price → reverse natural order
    private final TreeMap<BigDecimal, PriorityQueue<Order>> bids =
            new TreeMap<>(Comparator.reverseOrder());

    // Best ask = lowest price → natural ascending order
    private final TreeMap<BigDecimal, PriorityQueue<Order>> asks =
            new TreeMap<>();

    // O(1) lookup by orderId for cancellation
    private final Map<String, Order> orderIndex = new ConcurrentHashMap<>();

    private BigDecimal lastTradePrice = null;

    public OrderBook(String symbol) {
        this.symbol = symbol;
    }

    // ---------------------------------------------------------------
    //  MATCHING LOGIC
    // ---------------------------------------------------------------

    /**
     * Match an incoming aggressive order against resting orders.
     * Returns a list of trades generated during matching.
     *
     * Handles: MARKET, LIMIT, IOC, FOK
     */
    public List<Trade> match(Order incoming) {
        return switch (incoming.type()) {
            case MARKET   -> matchMarket(incoming);
            case LIMIT    -> matchLimit(incoming);
            case IOC      -> matchIoc(incoming);
            case FOK      -> matchFok(incoming);
            case STOP_LIMIT -> {
                // Stop-limit orders are held in a separate trigger map
                // and injected as LIMIT orders when stopPrice is touched.
                // For now, insert into book as a pending stop.
                yield List.of();
            }
        };
    }

    /**
     * Market order: consume opposite side greedily until filled or book empty.
     */
    private List<Trade> matchMarket(Order aggressor) {
        List<Trade> trades = new ArrayList<>();
        TreeMap<BigDecimal, PriorityQueue<Order>> oppositeSide =
                aggressor.isBuy() ? asks : bids;

        long remainingQty = aggressor.remainingQuantity();

        while (remainingQty > 0 && !oppositeSide.isEmpty()) {
            Map.Entry<BigDecimal, PriorityQueue<Order>> bestLevel =
                    oppositeSide.firstEntry();
            PriorityQueue<Order> levelQueue = bestLevel.getValue();

            while (remainingQty > 0 && !levelQueue.isEmpty()) {
                Order passive = levelQueue.peek();
                long fillQty = Math.min(remainingQty, passive.remainingQuantity());
                Trade trade = Trade.of(aggressor, passive, bestLevel.getKey(), fillQty);
                trades.add(trade);
                lastTradePrice = bestLevel.getKey();
                remainingQty -= fillQty;

                // Update passive order
                Order updatedPassive = passive.withRemainingQuantity(
                        passive.remainingQuantity() - fillQty);
                levelQueue.poll();  // remove old
                if (updatedPassive.remainingQuantity() > 0) {
                    levelQueue.offer(updatedPassive);  // put back if partial
                    orderIndex.put(updatedPassive.orderId(), updatedPassive);
                } else {
                    orderIndex.remove(passive.orderId());
                }
            }

            if (levelQueue.isEmpty()) {
                oppositeSide.remove(bestLevel.getKey());
            }
        }
        return trades;
    }

    /**
     * Limit order: match if crosses spread, otherwise insert into book.
     *
     * A BUY limit crosses if its price ≥ best ask.
     * A SELL limit crosses if its price ≤ best bid.
     */
    private List<Trade> matchLimit(Order aggressor) {
        List<Trade> trades = new ArrayList<>();
        TreeMap<BigDecimal, PriorityQueue<Order>> oppositeSide =
                aggressor.isBuy() ? asks : bids;

        long remainingQty = aggressor.remainingQuantity();

        while (remainingQty > 0 && !oppositeSide.isEmpty()) {
            Map.Entry<BigDecimal, PriorityQueue<Order>> bestLevel =
                    oppositeSide.firstEntry();
            BigDecimal bestPrice = bestLevel.getKey();

            // Check if the order crosses the spread
            boolean crosses = aggressor.isBuy()
                    ? aggressor.price().compareTo(bestPrice) >= 0
                    : aggressor.price().compareTo(bestPrice) <= 0;

            if (!crosses) break;  // no more crossable levels

            PriorityQueue<Order> levelQueue = bestLevel.getValue();
            while (remainingQty > 0 && !levelQueue.isEmpty()) {
                Order passive = levelQueue.peek();
                long fillQty = Math.min(remainingQty, passive.remainingQuantity());
                Trade trade = Trade.of(aggressor, passive, bestPrice, fillQty);
                trades.add(trade);
                lastTradePrice = bestPrice;
                remainingQty -= fillQty;

                Order updatedPassive = passive.withRemainingQuantity(
                        passive.remainingQuantity() - fillQty);
                levelQueue.poll();
                if (updatedPassive.remainingQuantity() > 0) {
                    levelQueue.offer(updatedPassive);
                    orderIndex.put(updatedPassive.orderId(), updatedPassive);
                } else {
                    orderIndex.remove(passive.orderId());
                }
            }
            if (levelQueue.isEmpty()) {
                oppositeSide.remove(bestPrice);
            }
        }

        // If not fully filled, rest the unfilled portion in the book
        if (remainingQty > 0) {
            Order restingOrder = aggressor.withRemainingQuantity(remainingQty);
            insertIntoBook(restingOrder);
        }

        return trades;
    }

    /**
     * IOC: match what's available, cancel the rest. Never rests in book.
     */
    private List<Trade> matchIoc(Order aggressor) {
        // Temporarily treat as LIMIT for matching, then discard remainder
        List<Trade> trades = matchLimit(aggressor);
        // Remove the resting portion we just inserted (if any)
        removeFromBook(aggressor.orderId());
        return trades;
    }

    /**
     * FOK: check if the full quantity can be filled before executing.
     * If not, reject entirely (zero trades).
     */
    private List<Trade> matchFok(Order aggressor) {
        long available = availableQuantityAtOrBetter(aggressor);
        if (available < aggressor.quantity()) {
            return List.of();  // Cannot fill completely — cancel
        }
        return matchMarket(aggressor);  // FOK matched as market once confirmed
    }

    // ---------------------------------------------------------------
    //  BOOK OPERATIONS
    // ---------------------------------------------------------------

    public void insertIntoBook(Order order) {
        TreeMap<BigDecimal, PriorityQueue<Order>> side =
                order.isBuy() ? bids : asks;
        side.computeIfAbsent(order.price(), p -> new PriorityQueue<>())
            .offer(order);
        orderIndex.put(order.orderId(), order);
    }

    public boolean removeFromBook(String orderId) {
        Order order = orderIndex.remove(orderId);
        if (order == null || order.price() == null) return false;

        TreeMap<BigDecimal, PriorityQueue<Order>> side =
                order.isBuy() ? bids : asks;
        PriorityQueue<Order> level = side.get(order.price());
        if (level == null) return false;

        boolean removed = level.removeIf(o -> o.orderId().equals(orderId));
        if (level.isEmpty()) side.remove(order.price());
        return removed;
    }

    // ---------------------------------------------------------------
    //  QUERIES
    // ---------------------------------------------------------------

    private long availableQuantityAtOrBetter(Order aggressor) {
        TreeMap<BigDecimal, PriorityQueue<Order>> oppositeSide =
                aggressor.isBuy() ? asks : bids;
        long total = 0;
        for (Map.Entry<BigDecimal, PriorityQueue<Order>> entry : oppositeSide.entrySet()) {
            BigDecimal levelPrice = entry.getKey();
            boolean crosses = aggressor.isBuy()
                    ? aggressor.price() == null || aggressor.price().compareTo(levelPrice) >= 0
                    : aggressor.price() == null || aggressor.price().compareTo(levelPrice) <= 0;
            if (!crosses) break;
            for (Order o : entry.getValue()) total += o.remainingQuantity();
        }
        return total;
    }

    public Optional<BigDecimal> bestBid() {
        return bids.isEmpty() ? Optional.empty() : Optional.of(bids.firstKey());
    }

    public Optional<BigDecimal> bestAsk() {
        return asks.isEmpty() ? Optional.empty() : Optional.of(asks.firstKey());
    }

    public Optional<BigDecimal> spread() {
        if (bids.isEmpty() || asks.isEmpty()) return Optional.empty();
        return Optional.of(asks.firstKey().subtract(bids.firstKey()));
    }

    public Optional<BigDecimal> lastTradePrice() {
        return Optional.ofNullable(lastTradePrice);
    }

    public int bidLevels() { return bids.size(); }
    public int askLevels() { return asks.size(); }
    public int totalOrders() { return orderIndex.size(); }
    public String symbol()   { return symbol; }

    /** Snapshot of top N bid/ask levels for WebSocket market data feed */
    public OrderBookSnapshot snapshot(int depth) {
        List<PriceLevel> bidLevels = new ArrayList<>();
        List<PriceLevel> askLevels = new ArrayList<>();

        int count = 0;
        for (Map.Entry<BigDecimal, PriorityQueue<Order>> e : bids.entrySet()) {
            if (count++ >= depth) break;
            long qty = e.getValue().stream().mapToLong(Order::remainingQuantity).sum();
            bidLevels.add(new PriceLevel(e.getKey(), qty, e.getValue().size()));
        }
        count = 0;
        for (Map.Entry<BigDecimal, PriorityQueue<Order>> e : asks.entrySet()) {
            if (count++ >= depth) break;
            long qty = e.getValue().stream().mapToLong(Order::remainingQuantity).sum();
            askLevels.add(new PriceLevel(e.getKey(), qty, e.getValue().size()));
        }
        return new OrderBookSnapshot(symbol, bidLevels, askLevels, lastTradePrice);
    }

    public record PriceLevel(BigDecimal price, long totalQuantity, int orderCount) {}
    public record OrderBookSnapshot(String symbol, List<PriceLevel> bids, List<PriceLevel> asks, BigDecimal lastTradePrice) {}
}

File: backend/src/engine/OrderBook.cpp
#include "titanx/engine/OrderBook.hpp"
#include <algorithm>

namespace titanx {

OrderBook::OrderBook(const std::string& symbol)
    : symbol_(symbol) {}

// ---------------------------------------------------------------
//  MAIN MATCHING ENTRY POINT
// ---------------------------------------------------------------

std::vector<Trade> OrderBook::match(const Order& incoming) {
    switch (incoming.type) {
        case OrderType::MARKET:     return matchMarket(incoming);
        case OrderType::LIMIT:      return matchLimit(incoming);
        case OrderType::IOC:        return matchIoc(incoming);
        case OrderType::FOK:        return matchFok(incoming);
        case OrderType::STOP_LIMIT:
            // Stop-limit orders are held in a separate trigger map
            // and injected as LIMIT orders when stopPrice is touched.
            return {};
    }
    return {};
}

// ---------------------------------------------------------------
//  MARKET ORDER: consume opposite side greedily
// ---------------------------------------------------------------

std::vector<Trade> OrderBook::matchMarket(const Order& aggressor) {
    std::vector<Trade> trades;

    // If buying, match against asks (sellers). If selling, match against bids.
    auto matchSide = [&](auto& oppositeSide) {
        int64_t remainingQty = aggressor.remainingQuantity;

        while (remainingQty > 0 && !oppositeSide.empty()) {
            auto bestIt = oppositeSide.begin();
            int64_t bestPrice = bestIt->first;
            auto& levelQueue = bestIt->second;

            while (remainingQty > 0 && !levelQueue.empty()) {
                Order& passive = levelQueue.front();
                int64_t fillQty = std::min(remainingQty, passive.remainingQuantity);

                Trade trade = Trade::create(aggressor, passive, bestPrice, fillQty);
                trades.push_back(trade);
                lastTradePrice_ = bestPrice;
                remainingQty -= fillQty;

                // Update passive order (copy-on-modify)
                Order updatedPassive = passive.withRemainingQuantity(
                    passive.remainingQuantity - fillQty);
                levelQueue.pop_front(); // remove old

                if (updatedPassive.remainingQuantity > 0) {
                    levelQueue.push_front(updatedPassive); // put back if partial
                    orderIndex_[updatedPassive.orderId] = updatedPassive;
                } else {
                    orderIndex_.erase(passive.orderId);
                }
            }

            if (levelQueue.empty()) {
                oppositeSide.erase(bestIt);
            }
        }
    };

    if (aggressor.isBuy()) {
        matchSide(asks_);
    } else {
        matchSide(bids_);
    }

    return trades;
}

// ---------------------------------------------------------------
//  LIMIT ORDER: match if crosses spread, else insert into book
// ---------------------------------------------------------------

std::vector<Trade> OrderBook::matchLimit(const Order& aggressor) {
    std::vector<Trade> trades;

    auto matchSide = [&](auto& oppositeSide) {
        int64_t remainingQty = aggressor.remainingQuantity;

        while (remainingQty > 0 && !oppositeSide.empty()) {
            auto bestIt = oppositeSide.begin();
            int64_t bestPrice = bestIt->first;

            // Check if the order crosses the spread
            bool crosses;
            if (aggressor.isBuy()) {
                crosses = aggressor.price >= bestPrice;
            } else {
                crosses = aggressor.price <= bestPrice;
            }

            if (!crosses) break; // no more crossable levels

            auto& levelQueue = bestIt->second;
            while (remainingQty > 0 && !levelQueue.empty()) {
                Order& passive = levelQueue.front();
                int64_t fillQty = std::min(remainingQty, passive.remainingQuantity);

                Trade trade = Trade::create(aggressor, passive, bestPrice, fillQty);
                trades.push_back(trade);
                lastTradePrice_ = bestPrice;
                remainingQty -= fillQty;

                Order updatedPassive = passive.withRemainingQuantity(
                    passive.remainingQuantity - fillQty);
                levelQueue.pop_front();

                if (updatedPassive.remainingQuantity > 0) {
                    levelQueue.push_front(updatedPassive);
                    orderIndex_[updatedPassive.orderId] = updatedPassive;
                } else {
                    orderIndex_.erase(passive.orderId);
                }
            }

            if (levelQueue.empty()) {
                oppositeSide.erase(bestIt);
            }
        }

        // If not fully filled, rest the unfilled portion in the book
        if (remainingQty > 0) {
            Order restingOrder = aggressor.withRemainingQuantity(remainingQty);
            insertIntoBook(restingOrder);
        }
    };

    if (aggressor.isBuy()) {
        matchSide(asks_);
    } else {
        matchSide(bids_);
    }

    return trades;
}

// ---------------------------------------------------------------
//  IOC: match what's available, cancel the rest
// ---------------------------------------------------------------

std::vector<Trade> OrderBook::matchIoc(const Order& aggressor) {
    std::vector<Trade> trades = matchLimit(aggressor);
    // Remove the resting portion we just inserted (if any)
    removeFromBook(aggressor.orderId);
    return trades;
}

// ---------------------------------------------------------------
//  FOK: check full quantity available before executing
// ---------------------------------------------------------------

std::vector<Trade> OrderBook::matchFok(const Order& aggressor) {
    int64_t available = availableQuantityAtOrBetter(aggressor);
    if (available < aggressor.quantity) {
        return {}; // Cannot fill completely — cancel
    }
    return matchMarket(aggressor); // FOK matched as market once confirmed
}

// ---------------------------------------------------------------
//  BOOK OPERATIONS
// ---------------------------------------------------------------

void OrderBook::insertIntoBook(const Order& order) {
    if (order.isBuy()) {
        bids_[order.price].push_back(order);
    } else {
        asks_[order.price].push_back(order);
    }
    orderIndex_[order.orderId] = order;
}

bool OrderBook::removeFromBook(const std::string& orderId) {
    auto it = orderIndex_.find(orderId);
    if (it == orderIndex_.end()) return false;

    Order order = it->second;
    orderIndex_.erase(it);

    if (order.price == 0) return false; // market orders have no price

    if (order.isBuy()) {
        auto levelIt = bids_.find(order.price);
        if (levelIt == bids_.end()) return false;
        auto& queue = levelIt->second;
        queue.erase(
            std::remove_if(queue.begin(), queue.end(),
                [&](const Order& o) { return o.orderId == orderId; }),
            queue.end());
        if (queue.empty()) bids_.erase(levelIt);
    } else {
        auto levelIt = asks_.find(order.price);
        if (levelIt == asks_.end()) return false;
        auto& queue = levelIt->second;
        queue.erase(
            std::remove_if(queue.begin(), queue.end(),
                [&](const Order& o) { return o.orderId == orderId; }),
            queue.end());
        if (queue.empty()) asks_.erase(levelIt);
    }

    return true;
}

// ---------------------------------------------------------------
//  QUERIES
// ---------------------------------------------------------------

int64_t OrderBook::availableQuantityAtOrBetter(const Order& aggressor) const {
    int64_t total = 0;

    auto countSide = [&](const auto& oppositeSide) {
        for (const auto& [levelPrice, queue] : oppositeSide) {
            bool crosses;
            if (aggressor.isBuy()) {
                crosses = aggressor.price == 0 || aggressor.price >= levelPrice;
            } else {
                crosses = aggressor.price == 0 || aggressor.price <= levelPrice;
            }
            if (!crosses) break;
            for (const auto& o : queue) {
                total += o.remainingQuantity;
            }
        }
    };

    if (aggressor.isBuy()) {
        countSide(asks_);
    } else {
        countSide(bids_);
    }

    return total;
}

std::optional<int64_t> OrderBook::bestBid() const {
    if (bids_.empty()) return std::nullopt;
    return bids_.begin()->first;
}

std::optional<int64_t> OrderBook::bestAsk() const {
    if (asks_.empty()) return std::nullopt;
    return asks_.begin()->first;
}

std::optional<int64_t> OrderBook::spread() const {
    auto bid = bestBid();
    auto ask = bestAsk();
    if (!bid || !ask) return std::nullopt;
    return *ask - *bid;
}

std::optional<int64_t> OrderBook::lastTradePriceValue() const {
    return lastTradePrice_;
}

OrderBookSnapshot OrderBook::snapshot(int depth) const {
    OrderBookSnapshot snap;
    snap.symbol = symbol_;
    snap.lastTradePrice = lastTradePrice_;

    int count = 0;
    for (const auto& [price, queue] : bids_) {
        if (count++ >= depth) break;
        int64_t qty = 0;
        for (const auto& o : queue) qty += o.remainingQuantity;
        snap.bids.push_back({price, qty, static_cast<int>(queue.size())});
    }

    count = 0;
    for (const auto& [price, queue] : asks_) {
        if (count++ >= depth) break;
        int64_t qty = 0;
        for (const auto& o : queue) qty += o.remainingQuantity;
        snap.asks.push_back({price, qty, static_cast<int>(queue.size())});
    }

    return snap;
}

} // namespace titanx
